// @sap/xssec/src/context/SecurityContext.js
const Types = require("../util/Types");
const { CORRELATIONID_HEADERS, FORWARDED_CLIENTCERTIFICATE_HEADER } = require('../util/constants');

/** 
 * @typedef {import('../service/Service')} Service
 * @typedef {import('../token/Token')} Token
 */

class SecurityContext {
    /** @type {Service} */
    service;
    /** @type {Token} */
    token;
    /** @type {Types.SecurityContextConfig} */
    config;

    constructor(service, token, contextConfig) {
        this.service = service;
        this.token = token;
        this.config = contextConfig;
    }    

    // Methods for backwards-compatibility   
    getAppToken() {
        return this.token.jwt;
    }

    getEmail() {
        return this.getUserInfo().email;
    }

    getExpirationDate() {
        return this.token.expirationDate;
    }

    getFamilyName() {
        return this.getUserInfo().familyName;
    }

    getGivenName() {
        return this.getUserInfo().givenName;
    }

    getGrantType() {
        return this.token.grantType;
    }

    getLogonName() {
        return this.getUserInfo().logonName
    }

    getUserInfo() {
        return {
            email: this.token.email,
            familyName: this.token.familyName,
            givenName: this.token.givenName,
            logonName: this.token.payload.user_name
        }
    }

    getTokenInfo() {
        return this.token;
    }

    /**
     * Tries to fill up missing properties of the security context configuration from the req object in the configuration.
     * @param {Types.SecurityContextConfig} contextConfig 
     */
    static buildContextConfig(contextConfig) {
        let { req } = contextConfig;

        for (let i = 0; contextConfig.correlationId == null && i < CORRELATIONID_HEADERS.length; i++) {
            contextConfig.correlationId = req?.headers?.[CORRELATIONID_HEADERS[i]];
        }
        contextConfig.clientCertificatePem ??= req?.headers?.[FORWARDED_CLIENTCERTIFICATE_HEADER];
        contextConfig.jwt ??= req?.headers?.authorization?.split(' ')[1];
    }
}


module.exports = SecurityContext;
