// @sap/xssec/src/util/constants.js
module.exports = {
    APP_TID_HEADER: "x-app_tid",
    AZP_HEADER: "x-azp",
    CLIENT_CERTIFICATE_HEADER : "x-client_cert",
    CLIENTID_HEADER: "x-client_id",
    CORRELATIONID_HEADER_VCAP: "x-vcap-request-id",
    CORRELATIONID_HEADERS: [
        "x-correlation-id",
        "x-request-id",
        "x-correlationid",
        "x-vcap-request-id"
    ],
    DEFAULT_TIMEOUT: 2000,
    FORWARDED_CLIENTCERTIFICATE_HEADER: "x-forwarded-client-cert",
    MAX_TIMEOUT: 5000,
    GRANTTYPE_CLIENTCREDENTIALS: "client_credentials",
    GRANTTYPE_JWTBEARER: "urn:ietf:params:oauth:grant-type:jwt-bearer",
    GRANTTYPE_PASSWORD: "password",
    HTTP_SCHEME : "http://",
    HTTPS_SCHEME : "https://",
    PEM_HEADER: "-----BEGIN CERTIFICATE-----",
    PEM_FOOTER: "-----END CERTIFICATE-----",
    USER_AGENT : "nodejs-xssec-4",
    SERVICE_PLAN_HEADER: "x-osb_plan",
    TOKEN_DATE_LEEWAY: 60, // seconds around nbf / exp time in which token is still considered valid
    X5T_CNF_CLAIM: "x5t#S256",
    ZID_HEADER: "x-zid",
    ZID_QUERY_PARAMETER: "zid",
}
